app/forms/concerns/validate_token_generation.rb

```diff
@@ -25,20 +26,27 @@ def ensure_time_between_requests
 
     return if time_since_last_mail > 2.minutes
 
-    errors[:base] << 'An email was sent less than two minutes ago. Please wait for a few minutes before trying again.'
+    errors[:base] <<
+      'An email was sent less than two minutes ago. Please wait for a few minutes before trying again.'
   end
 
   def email_should_not_have_bounced
     return if email.blank?
 
     return unless user&.email_bounced?
 
-    errors[:email] << "The email address you supplied cannot be used because an email we sent earlier bounced"
+    errors[:email] <<
+      'The email address you supplied cannot be used because an email we sent earlier bounced'
   end
 
   def user
-    @user ||= begin
-      current_school.users.with_email(email).first if email.present?
-    end
+    @user ||=
+      begin
+        current_school.users.with_email(email).first if email.present?
+      end
+  end
+
+  def token_generated_at
+    @token_generated_at ||= user&.login_token_generated_at
   end
```

Comment:
I noticed two blocking issues here:

1. The earlier implementation of this module did not have this method, even though it was referenced within module methods. This was because it was expected this method would be implemented in the _including_ class - where the token_generated_at would be different for the different use cases (reset password / login).
2. Using `login_token_generated_at` for determining when another email can be sent could be problematic. I haven't thought this through fully, but it seems to me that if a login token is generated by OAuth, that might block an email from being sent for two minutes..? Check for possible clashes.

Reply:
Yes, @harigopal Item 2 is possible. Can't think of a solution other than logging the email sent at as a separate column for this sign-in method as before.

Reply:
1. Fixed in https://github.com/pupilfirst/pupilfirst/pull/892/commits
2. Sharing a single column will result in such edge cases - https://github.com/pupilfirst/pupilfirst/issues/896